hudi-common/src/main/java/org/apache/hudi/common/table/view/AbstractTableFileSystemView.java

```diff
@@ -140,6 +141,22 @@ protected void init(HoodieTableMetaClient metaClient, HoodieTimeline visibleActi
    */
   protected void refreshTimeline(HoodieTimeline visibleActiveTimeline) {
     this.visibleCommitsAndCompactionTimeline = visibleActiveTimeline.getWriteTimeline();
+    this.timelineHashAndPendingReplaceInstants = null;
+  }
+
+  /**
+   * Get a list of pending replace instants. Caches the result for the active timeline.
+   * The cache is invalidated when {@link #refreshTimeline(HoodieTimeline)} is called.
+   *
+   * @return list of pending replace instant timestamps
+   */
+  private List<String> getPendingReplaceInstants() {
+    HoodieActiveTimeline activeTimeline = metaClient.getActiveTimeline();
```

Comment:
Does this code work for multi-thread? Guess the read lock would ensure the intergity, but anyway the variable assignment should be atomic.

Reply:
And do you think we should move the cache into the timeline itself, we already introduced some caches there for speeding up, check the `HoodieDefaultTimeline` for the cache.

Reply:
Regarding threading, should we just make this whole method synchronized?

Regarding caching, I'm open to whatever seems best. I noticed that the cache in HoodieDefaultTimeline is limited to a single instance and not a global cache. It seems like it may make sense long term to return the same instance whenever possible to benefit from this cache.

Reply:
> It seems like it may make sense long term to return the same instance whenever possible to benefit from this cache.

There should be no much difference because the map cache you use also has per-timeline granularity. The benefit to move to the timeline itself is for better maintainance.

Reply:
> > It seems like it may make sense long term to return the same instance whenever possible to benefit from this cache.
> 
> There should be no much difference because the map cache you use also has per-timeline granularity. The benefit to move to the timeline itself is for better maintainance.
> 
What do you mean by "map cache"?

> And if we move the cache inside the timeline, there should not be thread access conflicts.

Why is that? Can't multiple threads access the same timeline?

Reply:
> Can't multiple threads access the same timeline?

It could, and we should introduce some synchronized code guard for the access of the cache, we already did that for some caches in the timeline.

> What do you mean by "map cache"?

My typo, it's the "Pair" cache here.

Reply:
@danny0405 I looked at the `ActiveTimeline` class and the `DefaultTimeline` class but don't see any instances where there is a cache of instants for a particular action type. Can you point me in the right direction?

For this change, it is important that it is a `Set<String>` in the end so we don't need to keep recreating this set on each iteration.

Reply:
The cache is located in `HoodieDefaultTimeline`, both variable `instants` and `instantTimeSet` are lazy initialized cache.